### PR TITLE
Add Rockblock Satellite modem support

### PR DESCRIPTION
**Pull request by stephendade**
This is a (better) replacement for #21068, requiring far fewer modifications to the ArduPilot c code - "Less c, more Lua".

For bonus points, this also adds an example of encoding and decoding MAVLink packet in pure Lua.

This PR adds support for sending & receiving very basic MAVLink telemetry via a RockBlock SBD satellite modem.

It will allow for literal worldwide telemetry coverage for any vehicle fitted with the aforementioned modem.

The modem bandwidth is very low (~50 bytes per 20sec or so), so I've deliberately limited what can be sent and when (1 HIGH_LATENCY2 packet every 20 sec).

Requires https://github.com/stephendade/rockblock2mav at the GCS end to send and receive packets.

Tested with a real vehicle and modem.

### CODE REVIEW

**Comment by rmackay9** [comment]
I'm happy with the Copter arming changes. txs!

File: libraries/AP_Scripting/applets/RockBlock.lua [hunk 1]
```diff
@@ -0,0 +1,888 @@
...
+    -- private fields
+    local _mavbuffer = ""
+    local _mavresult = {}
+    local _payload_len = 0
+    local _mavdecodestate = 0 -- 0=looking for marker, 1=getting header,2=getting payload,3=getting crc
+    PROTOCOL_MARKER_V1 = 0xFE
+    HEADER_LEN_V1 = 6
+    local _txseqid = 0
+
+    -- AUTOGEN from MAVLink generator
```
Comment (magicrub) [inline libraries/AP_Scripting/applets/RockBlock.lua:57]: What is this for? Is it the CRC for individual packets? Do we care about all of these? Why not store just the ones we care about

File: libraries/AP_Scripting/applets/RockBlock.lua [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+hl2.eph = 0
+hl2.epv = 0
+hl2.temperature_air = 0
+hl2.climb_rate = 0
+hl2.battery = 0
+hl2.custom0 = 1 -- MAV_MODE_FLAG_CUSTOM_MODE_ENABLED
+hl2.custom1 = 0
+hl2.custom2 = 0
+
+function wrap_360(angle)
```
Comment (magicrub) [inline libraries/AP_Scripting/applets/RockBlock.lua:757]: Instead of reinventing the wheel, why not just expose all the C++ functions to scripting?

**Review by magicrub — COMMENTED** [review]
I don't think I understand the refactor. Why is there so much MAVLink handling in the scripts? I know you need to hijack the serial port for AT commands.. but I don't see why most of the MAVLink handling can't be handled with existing C/C++ code. Also, How does this impact Plane and Rover? Specifically, I'd love to see a Guided command work on Plane.

**Comment by stephendade** [comment]
> I don't think I understand the refactor. Why is there so much MAVLink handling in the scripts? I know you need to hijack the serial port for AT commands.. but I don't see why most of the MAVLink handling can't be handled with existing C/C++ code

Indeed, this is what the previous PR did. The issue was the required Lua bindings took up too much flash (2.3k) to justify putting the feature in.

**Comment by stephendade** [comment]
> Also, How does this impact Plane and Rover? Specifically, I'd love to see a Guided command work on Plane.

Tested on Plane, Rover and Copter. Guided command works on all 3.

File: libraries/AP_Scripting/applets/RockBlock.lua (hunk 1 above)
Comment (stephendade) [reply libraries/AP_Scripting/applets/RockBlock.lua:57]: Fixed - removed the CRC's we don't care about.

File: libraries/AP_Scripting/applets/RockBlock.lua [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+
+        -- send packet
+        rockblock.loadtxbuffer(newpkt)
+        
+    end
+
+    return HLSatcom, 100
+end
+
+return HLSatcom, 100
```
Comment (tridge) [inline libraries/AP_Scripting/applets/RockBlock.lua:751]: add a pcall wrapper so it restarts on an exception, see other applets for example

File: libraries/AP_Scripting/applets/RockBlock.lua [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+
+-- Caveats:
+-- This will *only* send HIGH_LATENCY2 packets via the SBD modem. No heartbeats,
+-- no command acknowledgements, no statustexts, no parameters, etc
+-- A single HIGH_LATENCY2 packet will be send every 20 sec
+-- MAVLink 1 will be used, as it's slightly more efficient (50 vs 52 bytes for a HL2 message)
+-- Any incoming packets on the first mailbox check will be ignored (as these may be from a long time in the past)
+-- Only 1 command can be sent at a time from the GCS. Any subsequent commands will overwrite the previous command
+-- The param SCR_VM_I_COUNT may need to be increased in some circumstances
+
```
Comment (tridge) [inline libraries/AP_Scripting/applets/RockBlock.lua:23]: a RCK_ENABLE param, and a param for frequency of send would be good

File: libraries/AP_Scripting/applets/RockBlock.lua [hunk 5]
```diff
@@ -0,0 +1,752 @@
+-- Lua script to send a recieve very basic MAVLink telemetry over a
```
Comment (tridge) [inline libraries/AP_Scripting/applets/RockBlock.lua:1]: use lua comment block

File: libraries/AP_Scripting/examples/MAVLinkHL.lua [hunk 6]
```diff
@@ -0,0 +1,457 @@
+-- Lua script to simulate a HL connection over a UART
```
Comment (tridge) [inline libraries/AP_Scripting/examples/MAVLinkHL.lua:1]: block comment

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by stephendade** [comment]
I've made the appropriate changes, based on @tridge's comments and tested again.

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
@stephendade nice work!

**Comment by magicrub** [comment]
@stephendade This is generating 16 warnings on the Lua docs autotest.
```
sudo apt install -y lua-check
./Tools/scripts/run_luacheck.sh
```
generates

```
Checking libraries/AP_Scripting/applets/RockBlock.lua 8 warnings
    libraries/AP_Scripting/applets/RockBlock.lua:116:5: mutating non-standard global variable _crc_extra
    libraries/AP_Scripting/applets/RockBlock.lua:117:5: mutating non-standard global variable _crc_extra
    libraries/AP_Scripting/applets/RockBlock.lua:118:5: mutating non-standard global variable _crc_extra
    libraries/AP_Scripting/applets/RockBlock.lua:119:5: mutating non-standard global variable _crc_extra
    libraries/AP_Scripting/applets/RockBlock.lua:122:5: mutating non-standard global variable _messages
    libraries/AP_Scripting/applets/RockBlock.lua:128:5: mutating non-standard global variable _messages
    libraries/AP_Scripting/applets/RockBlock.lua:134:5: mutating non-standard global variable _messages
    libraries/AP_Scripting/applets/RockBlock.lua:146:5: mutating non-standard global variable _messages

Checking libraries/AP_Scripting/examples/MAVLinkHL.lua 8 warnings
    libraries/AP_Scripting/examples/MAVLinkHL.lua:59:5: mutating non-standard global variable _crc_extra
    libraries/AP_Scripting/examples/MAVLinkHL.lua:60:5: mutating non-standard global variable _crc_extra
    libraries/AP_Scripting/examples/MAVLinkHL.lua:61:5: mutating non-standard global variable _crc_extra
    libraries/AP_Scripting/examples/MAVLinkHL.lua:62:5: mutating non-standard global variable _crc_extra
    libraries/AP_Scripting/examples/MAVLinkHL.lua:65:5: mutating non-standard global variable _messages
    libraries/AP_Scripting/examples/MAVLinkHL.lua:71:5: mutating non-standard global variable _messages
    libraries/AP_Scripting/examples/MAVLinkHL.lua:77:5: mutating non-standard global variable _messages
    libraries/AP_Scripting/examples/MAVLinkHL.lua:89:5: mutating non-standard global variable _messages

```

**Pull request merged**